### DISCUSSION TITLE
Sub-second data from Prometheus backend rounded to seconds

### DISCUSSION THREAD

**Discussion by martinjaeger** [grafana/grafana#44276] [Observability]
**What happened**:

We are collecting measurement data with sample rates down to 10 milliseconds. When using Prometheus backend (actually VictoriaMetrics, but with Prometheus API) the data is rounded to seconds, resulting in steps when displayed in a Time series panel.

The same data is displayed correctly when using InfluxDB as a backend.

Comparison of InfluxDB (top) and Prometheus (bottom) data sources:

![image](https://user-images.githubusercontent.com/17674105/148024103-b1fa82b0-c1d1-4a10-9009-bc94212bdf70.png)

**What you expected to happen**:

Prometheus supports 1 ms resolution, so the data should also be displayed with the same resolution.

I have verified that the scrape interval is set low enough:
![image](https://user-images.githubusercontent.com/17674105/148025010-f6242894-7593-4630-b275-a6092aee6db7.png)

**How to reproduce it (as minimally and precisely as possible)**:

Ingest data with sub-second time intervals to both databases and display the result.

**Anything else we need to know?**:

I've double-checked that the data read from the database itself has correct precision by observing HTTP network traffic with Wireshark. The Prometheus query results are fine.

I guess this is where the rounding (or in fact trimming) of the data happens:
https://github.com/grafana/grafana/blob/88bc47441f75b5956023be0b1f79c371420cdcdd/pkg/tsdb/prometheus/time_series_query.go#L306

This issue is somewhat related to the problem: #31948

While we're at it, maybe also OpenTSDB should get milliseconds support, as it seems to have the same type of rounding:
https://github.com/grafana/grafana/blob/88bc47441f75b5956023be0b1f79c371420cdcdd/pkg/tsdb/opentsdb/opentsdb.go#L182

**Environment**:
- Grafana version: v8.3.3 (30bb7a93ca)
- Data source type & version: InfluxDB v2.0 and Prometheus
- OS Grafana is installed on: Manjaro Linux
- User OS & Browser: Manjaro Linux, Chromium
- Grafana plugins: default

**Comment by ivanahuckova** [comment]
This wasn't possible even before Prometheus migration to backend data source and it is not possible in native Prometheus UI. However, this could be easy to implement by changing `timeField.Set(i, time.Unix(k.Timestamp.Unix(), 0).UTC())` to  `timeField.Set(i, time.Unix(0, k.Timestamp.UnixNano()).UTC())`.

**Comment by martinjaeger** [comment]
Thanks @ivanahuckova for your quick feedback. That looks like an easy fix, indeed.

I could create a quick PR for that, but I'm a bit reluctant to sign a CLA. Do you think someone of the Grafana team can take care of this feature request or will it take a while because it's obviously low priority?

**Comment by leeoniya** [comment]
@ivanahuckova this looks like the thing? https://github.com/grafana/grafana/pull/44707

**Reply by ivanahuckova** [reply]
Yes. I am not sure about all of the changes, I thought that we only need to change Prom code so the timestamp saved in dataFrame isn't rounded to seconds.

**Comment by ivanahuckova** [comment]
Fixed in https://github.com/grafana/grafana/pull/44707

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
